Add iceberg support to table_diff

williebsweet opened the pull request:
Resolves #4208

PR contains the following:
- Added a check of both source and target table formats 
- Added an error, should someone try to diff an Iceberg table and a Hive table
- Set the table format for a Iceberg/Iceberg diff to Iceberg
- Tests for various combinations of source/target/temp table formats

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/TobikoData/sqlmesh?pullRequest=4441) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/TobikoData/sqlmesh?pullRequest=4441) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/TobikoData/sqlmesh?pullRequest=4441) it.</sub>

izeigerman commented on sqlmesh/core/engine_adapter/base.py:
```diff
@@ -2203,6 +2203,19 @@ def temp_table(
             query_or_df, columns_to_types=columns_to_types, target_table=name
         )
 
+        if hasattr(self, "_query_table_type_or_raise") and hasattr(self, "source_table"):
```
Should this logic be in `athena.py` where these attributes are available?

erindru replied:
I... dont actually see how this works. Where does `self.source_table` come from? This isn't an engine adapter property.

table_diff [calls](https://github.com/TobikoData/sqlmesh/blob/529e52fa9d890dae3e56bb958e49f5788676e5dc/sqlmesh/core/table_diff.py#L434) `EngineAdapter.temp_table(...)` and passes the query that it's trying to store the results of, which looks something like:
```
with source as (select ... from source_table),
target as (select ... from target_table),
joined as (select ... from source full outer join target)
select * from joined
```

I actually think it's table_diff that should be supplying the table format, not the engine adapter introspecting the query to try and look up all the tables referenced in the query to see if any of them are Iceberg.

williebsweet replied:
@erindru Something more like [this previous commit](https://github.com/TobikoData/sqlmesh/pull/4441/commits/d45c73e487a5fc6c84f1cf7d6ea02b37822ead9a)?

erindru replied:
Yes, exactly. With perhaps a comment to show why there is an Athena-specific check and the addition of some tests

izeigerman commented:
@williebsweet thanks for the submission! Some tests will be helpful

williebsweet commented:
@erindru I moved the logic back into `table_diff`. I did the following:
- Added a check of both source and target table formats 
- Added an error, should someone try to diff an Iceberg table and a Hive table
- Set the table format for a Iceberg/Iceberg diff to Iceberg

@izeigerman I also added tests for the scenarios above.

williebsweet commented:
Forgot to mention - I also tested this with my company's SQLMesh project (installing the `--editable` package) and I was able to run a diff. 🎆

erindru commented on sqlmesh/core/table_diff.py:
```diff
@@ -431,7 +433,26 @@ def name(e: exp.Expression) -> str:
+                # Sets the temp table's format to Iceberg.
+                # If neither source nor target table is Iceberg, it defaults to Hive (Athena's default).
+                elif source_table_type == "iceberg" or target_table_type == "iceberg":
+                    raise SQLMeshError(
```
I _think_ this is right. Table diff relies on copying a sample of data from each table into a single table, which on most engines is fine because they have a consistent table format.

I know that timestamp types in particular are handled differently between Athena/Iceberg and Athena/Hive so I dont think we can say "if Iceberg is detected - make the diff table Iceberg" because copying Hive data into it verbatim can still cause an error

erindru commented:
Thanks @williebsweet ! This looks much better
